File: examples/vuln_example.py
"""
vuln_example.py
A realistic-looking Python file with several common vulnerabilities
that mirror the C vulnerabilities in vuln_example.c.
Used to demonstrate locate_vuln.py.
"""

import os
import sqlite3
import subprocess
import pickle
import hashlib

LOG_FILE = "/tmp/app.log"

# ------------------------------------------------------------------
# 1. Command injection — mirrors C's system(cmd) vuln
# ------------------------------------------------------------------
def run_report(filename):
    """Generate a report for the given filename."""
    os.system("cat " + filename)           # VULN: unsanitised filename → shell injection


# ------------------------------------------------------------------
# 2. SQL injection — mirrors C's unchecked string concat
# ------------------------------------------------------------------
def get_user(username):
    """Fetch a user record from the database."""
    conn = sqlite3.connect("users.db")
    cur = conn.cursor()
    query = "SELECT * FROM users WHERE name = '" + username + "'"  # VULN: SQL injection
    cur.execute(query)
    return cur.fetchall()


# ------------------------------------------------------------------
# 3. Arbitrary code execution via eval — mirrors C's format-string
# ------------------------------------------------------------------
def calculate(expression):
    """Evaluate a user-supplied mathematical expression."""
    return eval(expression)                # VULN: arbitrary code execution


# ------------------------------------------------------------------
# 4. Insecure deserialization — mirrors C's use-after-free (memory corruption)
# ------------------------------------------------------------------
def load_session(session_bytes):
    """Restore a user session from serialized bytes."""
    return pickle.loads(session_bytes)     # VULN: arbitrary object deserialization


# ------------------------------------------------------------------
# 5. Path traversal — mirrors C's integer-overflow / heap mis-allocation
# ------------------------------------------------------------------
def read_file(filename):
    """Read and return the contents of a file in the data directory."""
    path = "/var/app/data/" + filename     # VULN: no normalization, allows ../../../etc/passwd
    with open(path, "r") as f:
        return f.read()


# ------------------------------------------------------------------
# Safe helper — parameterised query (no vulnerability)
# ------------------------------------------------------------------
def safe_get_user(username):
    conn = sqlite3.connect("users.db")
    cur = conn.cursor()
    cur.execute("SELECT * FROM users WHERE name = ?", (username,))
    return cur.fetchall()


def log_message(msg):
    with open(LOG_FILE, "a") as f:
        f.write(msg + "\n")


if __name__ == "__main__":
    import sys
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <username>")
        sys.exit(1)
    username = sys.argv[1]
    print(get_user(username))
    run_report(username)
    log_message(username)

File: examples/vuln_example.c
/*
 * vuln_example.c
 * A realistic-looking C file with several common vulnerabilities.
 * Used to demonstrate locate_vuln.py.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define LOG_FILE "/tmp/app.log"

/* ------------------------------------------------------------------ */
/* 1. Buffer overflow — strcpy with no bounds check                    */
/* ------------------------------------------------------------------ */
void greet_user(char *username) {
    char buf[64];
    strcpy(buf, username);          /* VULN: no bounds check on username */
    printf("Hello, %s!\n", buf);
}

/* ------------------------------------------------------------------ */
/* 2. Format string vulnerability                                       */
/* ------------------------------------------------------------------ */
void log_message(char *msg) {
    FILE *f = fopen(LOG_FILE, "a");
    if (f) {
        fprintf(f, msg);            /* VULN: user-controlled format string */
        fclose(f);
    }
}

/* ------------------------------------------------------------------ */
/* 3. Command injection via system()                                   */
/* ------------------------------------------------------------------ */
void run_report(char *filename) {
    char cmd[256];
    snprintf(cmd, sizeof(cmd), "cat %s", filename);
    system(cmd);                    /* VULN: unsanitised filename injected into shell */
}

/* ------------------------------------------------------------------ */
/* 4. Use-after-free                                                   */
/* ------------------------------------------------------------------ */
void process_data(int *data, int len) {
    int *tmp = (int *)malloc(len * sizeof(int));
    memcpy(tmp, data, len * sizeof(int));
    free(tmp);
    printf("First element: %d\n", tmp[0]);  /* VULN: use-after-free */
}

/* ------------------------------------------------------------------ */
/* 5. Integer overflow leading to heap underallocation                 */
/* ------------------------------------------------------------------ */
void store_items(unsigned int count) {
    /* If count is close to UINT_MAX, count+1 wraps to 0 */
    int *items = (int *)malloc((count + 1) * sizeof(int));  /* VULN: integer overflow */
    if (!items) return;
    memset(items, 0, (count + 1) * sizeof(int));
    free(items);
}

/* ------------------------------------------------------------------ */
/* Safe helper — bounds-checked copy (no vulnerability)                */
/* ------------------------------------------------------------------ */
void safe_copy(char *dst, const char *src, size_t dst_size) {
    strncpy(dst, src, dst_size - 1);
    dst[dst_size - 1] = '\0';
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <username>\n", argv[0]);
        return 1;
    }
    greet_user(argv[1]);
    log_message(argv[1]);
    run_report(argv[1]);
    return 0;
}
